tidy3d/components/source.py

```diff
@@ -465,6 +465,29 @@ def _dir_vector(self) -> Tuple[float, float, float]:
         return None
 
 
+class BroadbandSource(Source, ABC):
+    """A source with frequency dependent field distributions."""
+
+    num_freqs: pydantic.NonNegativeInt = pydantic.Field(
+        1,
+        title="Number of Frequency Points",
+        description="Number of points to approximate the frequency dependence of injected field.",
+    )
+
+    @cached_property
+    def frequency_grid(self) -> np.ndarray:
+        """A Chebyshev grid used to approximate frequency dependence."""
+        fmin, fmax = self.source_time.frequency_range(4)
+        return 0.5 * (fmin + fmax) + 0.5 * (fmax - fmin) * np.cos(
+            0.5 * np.pi * (2 * np.flip(np.arange(self.num_freqs)) + 1) / self.num_freqs
+        )
+
+    def map_frequencies(self, freq_grid) -> np.ndarray:
+        """Map frequency values into (-1,1) interval."""
+        fmin, fmax = self.source_time.frequency_range(4)
+        return (freq_grid - 0.5 * (fmin + fmax)) / (0.5 * (fmax - fmin))
```

Comment:
Let's just move this to the backend; put it in `source.py` as a standalone function for now, after this is merged I will do the change of implementing `BackendSource` types and will include it there.